Sources/TensorFlow/Core/Utilities.swift

```diff
@@ -167,3 +167,167 @@ extension UnsafeMutablePointer where Pointee == CTensorHandle? {
+            0xca273eceea26619c, 0xd186b8c721c0c207, 0xeada7dd6cde0eb1e, 0xf57d4f7fee6ed178,
+            0x06f067aa72176fba, 0x0a637dc5a2c898a6, 0x113f9804bef90dae, 0x1b710b35131c471b,
+            0x28db77f523047d84, 0x32caab7b40c72493, 0x3c9ebe0a15c9bebc, 0x431d67c49c100d4c,
+            0x4cc5d4becb3e42b6, 0x597f299cfc657e2a, 0x5fcb6fab3ad6faec, 0x6c44198c4a475817]
+
+        var accumulated = self
+        let lengthInBits = accumulated.count * 8
+        let lengthBytes = lengthInBits.bytes(count: blockSize / 8)
+        
+        // Step 1: Append padding.
+        let msgLength = accumulated.count
+        // Append one bit(`UInt8` with one bit) to the message.
+        accumulated.append(0x80)
+        // Append `0` bits until the length of `accumulated` in bits is 448 (mod 512).
+        let max = blockSize * 7 / 8
+        accumulated += [UInt8](
+            repeating: 0,
+            count: msgLength % blockSize < max ?
+                max - 1 - (msgLength % blockSize) :
+                blockSize + max - 1 - (msgLength % blockSize))
+
+        // Step 2: Append the message length as a 64-bit representation of `lengthInBits`.
+        accumulated += lengthBytes
+
+        // Step 3: Process the array bytes.
+        var accumulatedHash = SIMD8<UInt64>(
+            0x6a09e667f3bcc908, 0xbb67ae8584caa73b, 0x3c6ef372fe94f82b, 0xa54ff53a5f1d36f1, 
+            0x510e527fade682d1, 0x9b05688c2b3e6c1f, 0x1f83d9abfb41bd6b, 0x5be0cd19137e2179)
+        var index = 0
+        while index < accumulated.count {
+            let chunk = accumulated[index..<(index + blockSize)]
+            index += blockSize
+            
+            // Break chunk into sixteen 64-bit words M[j], 0 ≤ j ≤ 15, in big-endian format.
+            // Extend the sixteen 64-bit words into eighty 64-bit words:
+            var M = [UInt64](repeating: 0, count: k.count)
+            for x in k.indices {
+                switch x {
+                case 0...15:
+                    let start = chunk.startIndex.advanced(by: x * 8)
+                    M[x] = UInt64(bytes: chunk, startingAt: start)
+                    break
+                default:
+                    let s0 = M[x - 15].rotate(rightBy: 1) ^ 
+                        M[x - 15].rotate(rightBy: 8) ^ 
+                        (M[x - 15] >> 7)
+                    let s1 = M[x - 2].rotate(rightBy: 19) ^ 
+                        M[x - 2].rotate(rightBy: 61) ^ 
+                        (M[x - 2] >> 6)
+                    M[x] = M[x - 16] &+ s0 &+ M[x - 7] &+ s1
+                    break
+                }
+            }
+
+            var A = accumulatedHash[0]
+            var B = accumulatedHash[1]
+            var C = accumulatedHash[2]
+            var D = accumulatedHash[3]
+            var E = accumulatedHash[4]
+            var F = accumulatedHash[5]
+            var G = accumulatedHash[6]
+            var H = accumulatedHash[7]
+
+            for j in 0..<k.count {
```

Comment:
I'm not sure if rotation is directly supported by SIMD APIs yet, but it'd be nice to vectorize this as well. (see [proposal](https://github.com/apple/swift-evolution/blob/master/proposals/0251-simd-additions.md)).

Reply:
That would be nice indeed. I believe there must be SIMD implementations of SHA512 out there. I just went for something simple to add support without dependencies but I'm sure we can make it much more efficient.

Reply:
@stephentyrone Could you please advise us?